brilirs/src/interp.rs

```diff
@@ -481,13 +559,35 @@ fn execute_value_op<T: std::io::Write>(
     Call => {
       let callee_func = state.prog.get(funcs[0]).unwrap();
 
-      make_func_args(callee_func, args, &mut state.env);
+      if callee_func.is_promise() {
+        let mut cloned = state.clone();
+        make_func_args(callee_func, args, &mut cloned.env);
 
-      let result = execute(state, callee_func)?.unwrap();
+        let callee_func = funcs[0];
+        let out = out.clone();
 
-      state.env.pop_frame();
+        let thread_stop_flag = Arc::new(AtomicBool::new(false));
+        let cloned_stop_flag = thread_stop_flag.clone();
+        handlers.insert(
+          dest,
+          (
+            thread::spawn(move || execute_thread(cloned, callee_func, out, cloned_stop_flag)),
+            thread_stop_flag,
+          ),
+        );
+      } else {
+        make_func_args(callee_func, args, &mut state.env);
+
+        let result = execute(state, funcs[0].clone(), out, stop_flag)?;
+
+        let val = match result {
+          Some(val) => val,
+          None => Value::Uninitialized,
```

Comment:
This result was originally an unwrap which I think better aligns with the program semantics. Typechecking should ensure that there is always a value here so it is not possible to create an Uninitialized value here